Support of multimeasure data sent from IOT device to iotagent-json

Chandradeep-NEC opened the pull request:
Resolving Issue #391

fgalan commented:
Thanks for the contribution!

Travis reports errors in lint stage. Please have a look and fix it.

In addition:

* PR should include a line in CHANGES_NEXT_RELEASE about the bugfix. Have a look to existing content in that file to get an idea of the style to use.
* What about unit tests? If functionality changes that should have an impact on test (we will see that once lint issues get fixes and travis can execute the unit test stage). If there is no impact, then new tests should be added to cover this fix.

chandradeep03 commented:
As per my observation, The UT error reported by Travis are not related with the PR #396.

In additional, can you please guide me as where do i need to add CHANGES_NEXT_RELEASE.
Please do  level this as a feature but not a bug.

chandradeep03 commented:
@fgalan  
Hi,
I've added the UT for #396 .
Can you please check and confirm it.
Thanks in advance.

fgalan commented:
Thanks for the feedback. I'd have a look as soon as I can, sorry...

fgalan commented:
For the sake of simplicity I'd suggest to merge https://github.com/chandradeep03/iotagent-json/pull/3 as it is now, then continue the review in this PR (it is easier that way than having to review at "two stages"). If new changes are required after that then you can merge your new PRs addressing it on chandradeep03:master directly and I'll see the result in this one (PR #396).

Does it make sense?

chandradeep03 commented:
> For the sake of simplicity I'd suggest to merge [chandradeep03#3](https://github.com/chandradeep03/iotagent-json/pull/3) as it is now, then continue the review in this PR (it is easier that way than having to review at "two stages"). If new changes are required after that then you can merge your new PRs addressing it on chandradeep03:master directly and I'll see the result in this one (PR #396).
> 
> Does it make sense?

Thanks for your response.

i've some confusion in your comment as per my understanding, the UT written in https://github.com/chandradeep03/iotagent-json/pull/3 should be the part of  https://github.com/telefonicaid/iotagent-json/pull/396. Please confirm.

How can i merge https://github.com/chandradeep03/iotagent-json/pull/3 in master as this PR is still opened.

fgalan commented:
> How can i merge chandradeep03#3 in master as this PR is still opened.

1. Go to https://github.com/chandradeep03/iotagent-json/pull/3
2. Look to the "Merge" button, at the bottom
3. Click it :)

chandradeep03 commented:
> > How can i merge [chandradeep03#3](https://github.com/chandradeep03/iotagent-json/pull/3) in master as this PR is still opened.
> 
> 1. Go to [chandradeep03#3](https://github.com/chandradeep03/iotagent-json/pull/3)
> 2. Look to the "Merge" button, at the bottom
> 3. Click it :)

Thanks for your guidance.
My concern regarding merging was as after merging the PR will still be open so how will you consider this open PR?
However i've merged [chandradeep03#3](https://github.com/chandradeep03/iotagent-json/pull/3) in master.

fgalan commented on test/unit/HTTP_reveice_measures-test_multiple_measurement.js:
```diff
@@ -0,0 +1,147 @@
+        var optionsMeasure = {
+            url: 'http://localhost:' + config.http.port + '/iot/json',
+            method: 'POST',
+            json:[ {
```
How many requests are sent to CB when a multimeasure like this is processed? One request per measure (4 requests in this case)? Or a single request with a "sum" of all the measures?

chandradeep03 replied:
Multi measure means one request will have multiple measurement data. SUch fucntionality is already implemented in iotagent-ul.
Here request is 1 but we are sending more than one measure (it can be 4 measure in this case).

fgalan replied:
Ok, one request.

Next question is, what is sent to CB when the same measure-name overlaps in different measures. In this case, humidity and temperature overlap as they appear in the four measures. I understand that the last one is the one send to CB.

In this case, the last one is humidity 35 and temperature 90 so I understand that CB will receive:

```

  "contextElements": [
    {
      "type": "AnMQTTDevice",
      "isPattern": "false",
      "id": "Second MQTT Device",
      "attributes": [
        {
          "name": "humidity",
          "type": "degrees",
          "value": "35"
        },
        {
          "name": "temperature",
          "type": "celsius",
          "value": "90"
        }
      ]
    }
  ],
  "updateAction": "UPDATE"
}
```

Is my interpretation correct?

chandradeep03 replied:
> Ok, one request.
> 
> Next question is, what is sent to CB when the same measure-name overlaps in different measures. In this case, humidity and temperature overlap as they appear in the four measures. I understand that the last one is the one send to CB.
> 
> In this case, the last one is humidity 35 and temperature 90 so I understand that CB will receive:
> 
> ```
> 
>   "contextElements": [
>     {
>       "type": "AnMQTTDevice",
>       "isPattern": "false",
>       "id": "Second MQTT Device",
>       "attributes": [
>         {
>           "name": "humidity",
>           "type": "degrees",
>           "value": "35"
>         },
>         {
>           "name": "temperature",
>           "type": "celsius",
>           "value": "90"
>         }
>       ]
>     }
>   ],
>   "updateAction": "UPDATE"
> }
> ```
> 
> Is my interpretation correct?

Yes your understanding correct.

fgalan replied:
However looking to the mocks expectations at 

```
.post('/v1/updateContext', utils.readExampleFile('./test/contextRequests/multipleMeasures.json'))
```

The [multipleMeasures.json](https://github.com/telefonicaid/iotagent-json/blob/master/test/contextRequests/multipleMeasures.json) is:

```
{
  "contextElements": [
    {
      "type": "AnMQTTDevice",
      "isPattern": "false",
      "id": "Second MQTT Device",
      "attributes": [
        {
          "name": "humidity",
          "type": "degrees",
          "value": "32"
        },
        {
          "name": "temperature",
          "type": "celsius",
          "value": "87"
        }
      ]
    }
  ],
  "updateAction": "UPDATE"
}
```

That is humidity=32 and temperature=87 and not humidity=35 and temperature=90 as it should be.

Thus, I understand something is not working as expected... What do you think?

chandradeep03 replied:
> However looking to the mocks expectations at
> 
> ```
> .post('/v1/updateContext', utils.readExampleFile('./test/contextRequests/multipleMeasures.json'))
> ```
> 
> The [multipleMeasures.json](https://github.com/telefonicaid/iotagent-json/blob/master/test/contextRequests/multipleMeasures.json) is:
> 
> ```
> {
>   "contextElements": [
>     {
>       "type": "AnMQTTDevice",
>       "isPattern": "false",
>       "id": "Second MQTT Device",
>       "attributes": [
>         {
>           "name": "humidity",
>           "type": "degrees",
>           "value": "32"
>         },
>         {
>           "name": "temperature",
>           "type": "celsius",
>           "value": "87"
>         }
>       ]
>     }
>   ],
>   "updateAction": "UPDATE"
> }
> ```
> 
> That is humidity=32 and temperature=87 and not humidity=35 and temperature=90 as it should be.
> 
> Thus, I understand something is not working as expected... What do you think?

The last measurement is the my expected value to update in CB. however i've again verified IotAgent Json with below data:

**Sensor data passed to IOT:**
[
{
"temperature": "50",
"humidity": "30"
},
{
"temperature": "60",
"humidity": "40"
}
]

**Measurment shown on CB:**
[
    {
        "id": "multiMeasure",
        "type": "multiType",
        "TimeInstant": {
            "type": "ISO8601",
            "value": "2019-06-20T07:44:23.00Z",
            "metadata": {}
        },
        "humidity": {
            "type": "string",
            "value": "40",
            "metadata": {
                "TimeInstant": {
                    "type": "ISO8601",
                    "value": "2019-06-20T07:44:23.750Z"
                }
            }
        },
        "temperature": {
            "type": "string",
            "value": "60",
            "metadata": {
                "TimeInstant": {
                    "type": "ISO8601",
                    "value": "2019-06-20T07:44:23.750Z"
                }
            }
        }
    }
]

fgalan replied:
Maybe we are not understanding each other :) I'll try again, from the beginning.

This unit test case models a situation in which you send a multi-measures with 4 items to the IOTA and check the request that CB will receive in that case. Not with a real CB but with a mock to check request expectations (I understand you are familiar with the general concepts of unit testing and mocks, otherwise please tell me).

As far as I understand by the previous conversation (maybe I have interpreted something incorrectly, in that case, please tell me), when the IOTA receives a multi-measure with N elements all them are "summed together" so only 1 request is sent to the CB. Moreover, if the same attribute appears in several measures (as humidity and temperature in this case) the one corresponding to the last measure is the one which overrides the others. In the particular example of this test the last measure is humidity: '35' and temperature: '90'.

Thus CB should receive a request with humidity: '35' and temperature: '90', shouldn't it?

**However**, let's analyze the expectations that are being used in the mock. That's the line:

```
.post('/v1/updateContext', utils.readExampleFile('./test/contextRequests/multipleMeasures.json'))
```

and if we look to that test/contextRequests/multipleMeasures.json file in the repository we have:

```
{
  "contextElements": [
    {
      "type": "AnMQTTDevice",
      "isPattern": "false",
      "id": "Second MQTT Device",
      "attributes": [
        {
          "name": "humidity",
          "type": "degrees",
          "value": "32"
        },
        {
          "name": "temperature",
          "type": "celsius",
          "value": "87"
        }
      ]
    }
  ],
  "updateAction": "UPDATE"
}
```

That is temperature 87 and humidity 32. And 35 != 32 and 90 != 87. Thus, in summary, the test is not correct (in fact, I wonder how travis is passing this as ok...)

I hope to have clarified the problem... Please confirm you understand it (or tell what is wrong in my discussion, maybe I have missed something :).

chandradeep03 replied:
> Maybe we are not understanding each other :) I'll try again, from the beginning.
> 
> This unit test case models a situation in which you send a multi-measures with 4 items to the IOTA and check the request that CB will receive in that case. Not with a real CB but with a mock to check request expectations (I understand you are familiar with the general concepts of unit testing and mocks, otherwise please tell me).
> 
> As far as I understand by the previous conversation (maybe I have interpreted something incorrectly, in that case, please tell me), when the IOTA receives a multi-measure with N elements all them are "summed together" so only 1 request is sent to the CB. Moreover, if the same attribute appears in several measures (as humidity and temperature in this case) the one corresponding to the last measure is the one which overrides the others. In the particular example of this test the last measure is humidity: '35' and temperature: '90'.
> 
> Thus CB should receive a request with humidity: '35' and temperature: '90', shouldn't it?
> 
> **However**, let's analyze the expectations that are being used in the mock. That's the line:
> 
> ```
> .post('/v1/updateContext', utils.readExampleFile('./test/contextRequests/multipleMeasures.json'))
> ```
> 
> and if we look to that test/contextRequests/multipleMeasures.json file in the repository we have:
> 
> ```
> {
>   "contextElements": [
>     {
>       "type": "AnMQTTDevice",
>       "isPattern": "false",
>       "id": "Second MQTT Device",
>       "attributes": [
>         {
>           "name": "humidity",
>           "type": "degrees",
>           "value": "32"
>         },
>         {
>           "name": "temperature",
>           "type": "celsius",
>           "value": "87"
>         }
>       ]
>     }
>   ],
>   "updateAction": "UPDATE"
> }
> ```
> 
> That is temperature 87 and humidity 32. And 35 != 32 and 90 != 87. Thus, in summary, the test is not correct (in fact, I wonder how travis is passing this as ok...)
> 
> I hope to have clarified the problem... Please confirm you understand it (or tell what is wrong in my discussion, maybe I have missed something :).

Thanks for the explanation.
I understood your point.
The mock object contextBrokerMock of beforeEach block written in "HTTP_reveice_measures-test_multiple_measurement.js" is not used in the execution of below test case
        it('should return a 200 OK with no error', function() {
            request(optionsMeasure, function(error, result, body) {
                should.not.exist(error);
                result.statusCode.should.equal(200);
                //done();
            });
        });
So the expectation of multipleMeasure.json is not as per request data provided in optionsMeasure.
In the UT it test only returning of no error and sttaus code as 200.
From UT  code, beforeEach block should be removed. I'll update this.
Please let me know if more explanation is required.

fgalan replied:
I didn't realize about the commented line `//done()` :O. Please, don't do that. The testing framework needs done() to be called after test termination. Not doing so can bring unexpected problems. This documentation may help to get the point: https://mochajs.org/#asynchronous-code

> The mock object contextBrokerMock of beforeEach block written in "HTTP_reveice_measures-test_multiple_measurement.js" is not used in the execution of below test case

This shoulf be fixed. The test has to check the expectaions of the mock who plays the role of CB. Otherwise, the test is pointless :)

In particular what I mean is:

1. Create a JSON file similar to multipleMeasures.json but modeling your expected case (humidity 35 and temperature 90). Name it for instance multipleMeasures2.json
2. Use that file in contextBrokerMock() and get the test passed
3. Check that if you change something in multipleMeasures2.json (e.g. humidity set to 30) your test fails

fgalan commented:
It seems that PR has some lint errors looking to Travis log:

```
lib/bindings/HTTPBinding.js: line 216, col 26, 'i' is already defined.
lib/bindings/HTTPBinding.js: line 315, col 71, Functions declared within loops referencing an outer scoped variable may lead to confusing semantics. (next, iotaUtils, req, arr, i, sendConfigurationToDevice, replyToDevice)
```

They have to be fixed, please.

fgalan commented:
Travis still fails after last commits (please see https://travis-ci.org/telefonicaid/iotagent-json/builds/545698081?utm_source=github_status&utm_medium=notification). I understand you keep on it.

Chandradeep-NEC commented:
@fgalan 
We have fix the UT and also some source files updated for failed test cases.
Sorry for delay in response and modification

fgalan commented:
> We have fix the UT and also some source files updated for failed test cases.
> Sorry for delay in response and modification

Thanks for the update! However, note that the date of the last commit in the PR at the present moment is June 18th. I understand that the fix of the UT you mention is pending to be pushed to the branch, isn't it?

fgalan commented:
...or maybe the date is when the commit was done but it hasn't been pushed until recently.

Please confirm if the fix has been pushed into the PR's branch so I can re-review again. Thanks!

Chandradeep-NEC commented:
@fgalan 
I've committed the modified source files and test files.
All UT have been passed successfully with Node8, Node10 and Node12 on my local environment.
Please check and review the modified files.

Chandradeep-NEC commented:
> @fgalan
> I've committed the modified source files and test files.
> All UT have been passed successfully with Node8, Node10 and Node12 on my local environment.
> Please check and review the modified files.

@fgalan 
Please provide your response.

Chandradeep-NEC commented:
@fgalan 
Can you please let me know if there is any update on this.

fgalan commented on test/unit/HTTP_reveice_measures-test2.js:
```diff
@@ -0,0 +1,371 @@
+        async.series([iotAgentLib.clearAll, iotagentMqtt.stop], done);
+    });
+
+    describe('When a POST measure arrives for the HTTP binding', function() {
```
```suggestion
    describe('When a POST multimeasure arrives for the HTTP binding', function() {
```

fgalan commented on test/unit/HTTP_reveice_measures-test2.js:
```diff
@@ -0,0 +1,371 @@
+        });
+    });
+
+    describe('When a POST measure arrives with a TimeInstant attribute in the body', function() {
```
```suggestion
    describe('When a POST multimeasure arrives with a TimeInstant attribute in the body', function() {
```

fgalan commented on test/unit/HTTP_reveice_measures-test2.js:
```diff
@@ -0,0 +1,371 @@
+            contextBrokerMock = nock('http://192.168.1.1:1026')
+                .matchHeader('fiware-service', 'smartGondor')
+                .matchHeader('fiware-servicepath', '/gardens')
+                .post('/v1/updateContext')
```
You should provide the expected payload for the /v1/updateContext operation here, similar to what you did in L128. Maybe you need to create a test/contextRequests/timeInstantMeasures2.json file for it.

fgalan commented on test/unit/HTTP_reveice_measures-test2.js:
```diff
@@ -0,0 +1,371 @@
+        });
+    });
+
+    describe('When a POST measure arrives with a TimeInstant query parameter in the body', function() {
```
```suggestion
    describe('When a POST multimeasure arrives with a TimeInstant query parameter in the body', function() {
```

fgalan commented on test/unit/HTTP_reveice_measures-test2.js:
```diff
@@ -0,0 +1,371 @@
+            contextBrokerMock = nock('http://192.168.1.1:1026')
+                .matchHeader('fiware-service', 'smartGondor')
+                .matchHeader('fiware-servicepath', '/gardens')
+                .post('/v1/updateContext')
```
Same than https://github.com/telefonicaid/iotagent-json/pull/396/files#r344287993

fgalan commented on test/unit/HTTP_reveice_measures-test2.js:
```diff
@@ -0,0 +1,371 @@
+        });
+    });
+
+    describe('When a POST measure arrives for an unprovisioned device', function() {
```
```suggestion
    describe('When a POST multimeasure arrives for an unprovisioned device', function() {
```

fgalan commented on test/unit/HTTP_reveice_measures-test2.js:
```diff
@@ -0,0 +1,371 @@
+            contextBrokerUnprovMock = nock('http://unexistentHost:1026')
+                .matchHeader('fiware-service', 'smartGondor')
+                .matchHeader('fiware-servicepath', '/gardens')
+                .post('/v1/updateContext')
```
Same than https://github.com/telefonicaid/iotagent-json/pull/396/files#r344287993

fgalan commented on test/unit/ngsiv2/HTTP_reveice_measures-test2.js:
```diff
@@ -0,0 +1,391 @@
+        async.series([iotAgentLib.clearAll, iotaJson.stop], done);
+    });
+
+    describe('When a POST measure arrives for the HTTP binding', function() {
```
```suggestion
    describe('When a POST multimeasure arrives for the HTTP binding', function() {
```

fgalan commented on test/unit/ngsiv2/HTTP_reveice_measures-test2.js:
```diff
@@ -0,0 +1,391 @@
+            contextBrokerMock = nock('http://192.168.1.1:1026')
+                .matchHeader('fiware-service', 'smartGondor')
+                .matchHeader('fiware-servicepath', '/gardens')
+                .post('/v2/entities?options=upsert')
```
You should provide the expected payload for the /v2/entities?options=upsert operation here, similar to what you did in L152. Maybe you need to create a /test/unit/ngsiv2/contextRequests/timeInstantMeasures2.json file for it.

fgalan commented on test/unit/ngsiv2/HTTP_reveice_measures-test2.js:
```diff
@@ -0,0 +1,391 @@
+            contextBrokerMock = nock('http://192.168.1.1:1026')
+                .matchHeader('fiware-service', 'smartGondor')
+                .matchHeader('fiware-servicepath', '/gardens')
+                .post('/v2/entities?options=upsert')
```
Similar comment to https://github.com/telefonicaid/iotagent-json/pull/396/files#r344290893

fgalan commented on test/unit/ngsiv2/HTTP_reveice_measures-test2.js:
```diff
@@ -0,0 +1,391 @@
+        });
+    });
+
+    describe('When a POST measure arrives with a TimeInstant attribute in the body', function() {
```
```suggestion
    describe('When a POST multimeasure arrives with a TimeInstant attribute in the body', function() {
```

fgalan commented on test/unit/ngsiv2/HTTP_reveice_measures-test2.js:
```diff
@@ -0,0 +1,391 @@
+        });
+    });
+
+    describe('When a POST measure arrives with a TimeInstant query parameter in the body', function() {
```
```suggestion
    describe('When a POST multimeasure arrives with a TimeInstant query parameter in the body', function() {
```

fgalan commented on test/unit/ngsiv2/HTTP_reveice_measures-test2.js:
```diff
@@ -0,0 +1,391 @@
+        });
+    });
+
+    describe('When a POST measure arrives for an unprovisioned device', function() {
```
```suggestion
    describe('When a POST multimeasure arrives for an unprovisioned device', function() {
```

fgalan commented on test/unit/ngsiv2/HTTP_reveice_measures-test2.js:
```diff
@@ -0,0 +1,391 @@
+            contextBrokerUnprovMock = nock('http://unexistentHost:1026')
+                .matchHeader('fiware-service', 'smartGondor')
+                .matchHeader('fiware-servicepath', '/gardens')
+                .post('/v2/entities?options=upsert')
```
Similar case to https://github.com/telefonicaid/iotagent-json/pull/396/files#r344290893

fgalan commented:
I have reviewed the new versoin of the PR. Thanks for the contribution! I think it has improved since last time I saw and you are in the good track. I have added some additional comments.

In addition, given that the iota-json agent has evolved quite a bit since your last commits in this PR, please upgrade your PR brach with last master branch version (i.e. merging master branch into you PR branch).

What about other transports apart from HTTP (in particular I refer to MQTT and AMQP)? Does the multimeasure functionality works for them in this PR? Or they should be developed in a next PR as future development?

Finally, don't forget the PR should include a line in CHANGES_NEXT_RELEASE about the bugfix. Have a look to existing content in that file to get an idea of the style to use.

Chandradeep-NEC commented:
> I have reviewed the new version of the PR. Thanks for the contribution! I think it has improved since last time I saw and you are in the good track. I have added some additional comments.
> 
> In addition, given that the iota-json agent has evolved quite a bit since your last commits in this PR, please upgrade your PR brach with last master branch version (i.e. merging master branch into you PR branch).
> 
> What about other transports apart from HTTP (in particular I refer to MQTT and AMQP)? Does the multimeasure functionality works for them in this PR? Or they should be developed in a next PR as future development?
> 
> Finally, don't forget the PR should include a line in CHANGES_NEXT_RELEASE about the bugfix. Have a look to existing content in that file to get an idea of the style to use.

@fgalan 
We have updated the PR which includes incorporation of all comments as well as rebasing of branch. This PR supports multimeasure functionality only for HTTP. This feature will be developed for other transport  i.e. MQTT and AMQP in future as next PR.
CNR has been updated for this bug fix.

Chandradeep-NEC commented:
Hi @fgalan 
Can you please provide your response. If the modification is ok then this PR can be merged.

fgalan commented:
> Hi @fgalan
> Can you please provide your response. If the modification is ok then this PR can be merged.

I'll try to do it by the next week. Sorry for the delay.

In the meanwhile it would be great if you could uptrade your PR branch with master, as it seems to be some commits behind:

![imagen](https://user-images.githubusercontent.com/1534240/70257642-f9acd280-178a-11ea-88d7-ccc21b4782f1.png)

and, in addition, CHANGES_NEXT_RELEASE shows a conflict.

Thanks!

fgalan commented:
The PR seems to be ok. Please solve the conflict in CHANGEX_NEXT_RELEASE, upgrade with the latest from master so we can progress it.

Chandradeep-NEC commented:
Hi @fgalan,
I have resolved the conflict in CHANGES_NEXT_RELEASE and updated the branch as well.
Please merge the PR into master if it is OK.

Chandradeep-NEC commented:
@fgalan , @AlvaroVega 
Thanks for your support.
As the both PR #396 and #459 have been merged then can we close this issue #391 ?

Thanks :)

fgalan commented:
Issue #391 cannot be closed yet, as there are pending works:

> What is pending after this couple of PR: implement multimeasures for MQTT and AMPQ transports (and modifiy the IMPORTANT NOTE in documentation accordingly)

Chandradeep-NEC commented:
> Issue #391 cannot be closed yet, as there are pending works:
> 
> What is pending after this couple of PR: implement multimeasures for MQTT and AMPQ transports (and modifiy the IMPORTANT NOTE in documentation accordingly)

I understood your point. But I suggest to create another issue for implementation of multimeasure support of MQTT and AMPQ. We can provide reference of #391 and #396 to new issue.

The idea of creation of new issue came because the #391 has been opened from many days and also long discussion has been done on this. It would be good to add final conclusion points of #391 on new issue and it will prevent in any kind of confusion for developer.

Thanks :)

fgalan commented:
No problem. Let's do that.
